### PR TITLE
Add support for setting the group on a unix domain socket

### LINKED ISSUE #873: [NEW] Add support for setting the group on a unix domain socket
From a mastodon conversation.
```
By default the socket is created in the wheel group, which I clearly don’t want to add my application user use to, so I checked if I can change the group somehow. Chown worked, but got lost after a reboot.

It just felt like it was missing, both Postgres and php-fpm, which I configured right before, had an easy option to set the group to make exactly this possible.
Or am I missing something about how this should be done?
```

This seems fairly straightforward to do. We already have a [permission config](https://github.com/valkey-io/valkey/blob/0fc43edc6cf388854bc824f192647cc59f7019e7/src/config.c#L3204). So we just all need to support calling `chgrp()` here https://github.com/valkey-io/valkey/blob/0fc43edc6cf388854bc824f192647cc59f7019e7/src/anet.c#L515. It shouldn't be a difficult change, but will require a lot of code updating.

Can confirm it's also part of the postgres configuration: https://www.postgresql.org/docs/current/runtime-config-connection.html#GUC-UNIX-SOCKET-GROUP.

### PR DESCRIPTION
**Pull request by ayush933**
Add new optional, immutable string config called `unixsocketgroup`. 
Change the group of the unix socket to `unixsocketgroup` after `bind()` if specified.

Manually verified that group is modified as expected and fails on illegal arguments (like non existent group name).

Fixes #873.

### CODE REVIEW

**Comment by ayush933** [comment]
Hi, @madolson can you please take a look?

File: src/server.c
Comment (madolson) [inline src/server.c]:
```suggestion
        listener->group = server.unixsocketgroup;
```
If it's not set it's NULL, so it should be okay to always just set this. I prefer this so it's clearer what the value is.

File: src/anet.c
Comment (madolson) [inline src/anet.c]:
```suggestion
        if (chown(((struct sockaddr_un *)sa)->sun_path, /*owner remains same*/ -1, grp->gr_gid) == -1) {
```
Is the documented return value.

**Review by madolson — COMMENTED** [review]
It would also be nice to write a test to validate this, probably in [other](https://github.com/valkey-io/valkey/blob/unstable/tests/unit/other.tcl#L410) or misc. We can validate both the permissions and group.

File: src/connection.h
Comment (enjoy-binbin) [inline src/connection.h]: i see we put unixsocketperm into priv, and here we are adding a new group field? i suppose it is ok since the code is easy to write/read.

File: src/anet.c
Comment (enjoy-binbin) [reply src/anet.c]: i don't like putting the comment inline into the code, lets start it in a newline.

File: src/connection.h
Comment (madolson) [reply src/connection.h]: I was also unsure about this. The other idea I had was to make it priv1/priv2 to just keep it simple. I don't really want to allocate extra memory just to put in a private field.

Comment (enjoy-binbin) [reply src/connection.h]: > I don't really want to allocate extra memory just to put in a private field.

yes, i agree. let's use priv1/priv2

**Comment by ayush933** [comment]
Addressed reviews and added UT to test that permissions are set correctly.
Not sure which group to use in tests since using the default root group or creating new group would require sudo.

File: tests/unit/other.tcl
Comment (madolson) [inline tests/unit/other.tcl]:
```suggestion
set group [lindex [exec groups [exec whoami]] 0]
if {group != ""} {
    start_server [list tags {"repl external:skip"} overrides [list {unixsocketperm} {744} {unixsocketgroup} $group]] {
        test {test unixsocket options are set correctly} {
            set socketpath [lindex  [r config get unixsocket] 1]
            set attributes [file attributes $socketpath]
            set permissions [string range [dict get $attributes -permissions] end-2 end]
            assert_equal 744 $permissions
        }
    }
}
```
something like this might work.

Comment (ayush933) [reply tests/unit/other.tcl]: So the test were failing on ubuntu using this method with error
```
unixsocket /home/runner/work/valkey/valkey/tests/tmp/server.3167.13/socket
unixsocketgroup runner
.
.
8029:M 17 Aug 2024 17:25:19.958 # Failed opening Unix socket: chown: Operation not permitted
```
Updated the test to use the group from  a temp file instead. 
Can you please check and re-run the tests?

File: src/anet.c
Comment (madolson) [inline src/anet.c]:
```suggestion
        if ((grp = getgrnam(group)) == NULL) {
            anetSetError(err, "getgrnam error for group '%s': %s", group, strerror(errno));
            close(s);
            return ANET_ERR;
        }

        /* Owner of the socket remains same. */
        if (chown(((struct sockaddr_un *)sa)->sun_path, -1, grp->gr_gid) == -1) {
            anetSetError(err, "chown error for group '%s': %s", group, strerror(errno));
            close(s);
            return ANET_ERR;
        }
```
Based on the failure, was also going to suggest we add a little bit more debugging info.

File: tests/unit/other.tcl
Comment (ayush933) [reply tests/unit/other.tcl]: Marking as resolved since all the tests passed.

File: src/anet.c
Comment (ayush933) [reply src/anet.c]: Done.

**Comment by madolson** [comment]
@valkey-io/core-team Please vote on this small feature, we can optionally include it in Valkey 8.0 if it's ready for rc2, but not after.

File: src/connection.h
Comment (enjoy-binbin) [inline src/connection.h]:
```suggestion
    void *priv2; /* used by connection type specified data */
```

File: tests/unit/other.tcl
Comment (enjoy-binbin) [inline tests/unit/other.tcl]:
```suggestion
            set socketpath [lindex [r config get unixsocket] 1]
```

**Review by enjoy-binbin — APPROVED** [review]

File: src/connection.h
Comment (zuiderkwast) [reply src/connection.h]: *Two* custom type-specific fields? It looks like a hack. It makes the API a bit messy. Can't the unix socket listener embed both in the same priv field, for example in another allocated struct with two fields (perm and group)?

I don't feel strongly about it though.

Comment (enjoy-binbin) [reply src/connection.h]: see https://github.com/valkey-io/valkey/pull/901#pullrequestreview-2237090821

Comment (zuiderkwast) [reply src/connection.h]: OK!

File: src/config.c
Comment (hwware) [inline src/config.c:3113]: If the default group is wheel, I prefer setting it here.

File: src/connection.h
Comment (hwware) [inline src/connection.h:148]: IMO, I do not like name the variable name as priv1, priv2.  Later, we will not know the meaning they represent.  I believe priv1 represents unixsocketperm, and thus maybe void *group or ugroup is better than now to represent unixsocketgroup

**Review by madolson — APPROVED** [review]
LGTM, just waiting on majority approval

File: src/config.c
Comment (ayush933) [reply src/config.c:3113]: wheel is a privileged group on most systems and not the default.

File: src/connection.h
Comment (ayush933) [reply src/connection.h:148]: see https://github.com/valkey-io/valkey/pull/901#pullrequestreview-2237090821

**Comment by madolson** [comment]
Since the new test is a bit odd, running the daily against a bunch of different platforms. https://github.com/valkey-io/valkey/actions/runs/10530596918

**Review by lilpoozie2005 — APPROVED** [review]

**Pull request merged**
